meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java

```diff
@@ -75,8 +133,19 @@ public Permit acquire() {
    * @param count number of permits to acquire
    * @return a collection of permits
    */
-  public Collection<Permit> acquire(int count) {
-    throw new NotImplementedException("Not implemented yet.");
+  public synchronized Collection<Permit> acquire(int count) {
+    if (getRemainingCapacity() < count) {
+      LOG.warn("No sufficient permits available. Attempt to acquire {} permits, but only {} permits available", count,
+          getRemainingCapacity());
+      return null;
+    } else {
+      updateAcquirePermit(count);
```

Comment:
Event with `synchronized`, it is still possible that when we call `getRemainingCapacity`, there are enough permits but there is no permits when `updateAcquirePermit` actually get called. I think we need transaction here.

Reply:
After an offline discussion we agreed that it was necessary to implement some sort of anti race condition that prevents two clients from modifying a semaphore at the same time / in between processes. We came up with multiple solutions but ultimately agreed that we would check for the available capacity in the update call rather than having it separately.

Reply:
I'm still trying to understand how current implementation is atomic? We check and update znode with remaining capacity in updateAcquirePermit (nit :i feel it should be renamed to checkAcquistionAndUpdatePermit or something) and then update in-memory object in retrievePermit() but it's not atomic right? There still could be race conditions here. For ex there are two acquire calls with different numbers then caller may get different values based on order of execution.

Reply:
Could you please elaborate a bit more on "caller may get different values". What does `value` refer to?

Reply:
I meant different count values for acquire.

Reply:
We had an discussion offline. The client won't store the number of left permit in memory. Client uses `update` to update remaining permit on ZK, so there won't be race condition.